dask_match/datasets.py

```diff
@@ -37,17 +47,25 @@ def _divisions(self):
 
     @functools.cached_property
     def random_state(self):
-        return np.random.randint(2e9, size=self.npartitions)
+        size = (
+            len(pd.date_range(start=self.start, end=self.end, freq=self.partition_freq))
+            + 1
+        )
+        return np.random.randint(2e9, size=size)
 
     def _task(self, index):
+        original_index = index
+        if self._take_partitions is not None:
+            # Need original index for random_state
+            original_index = self._take_partitions[index]
```

Comment:
I propose that we handle this in `Expr._layer` instead

Reply:
Maybe something like this?

```python
class Expr:
    @property
    def _partitions(self):
        return range(self.npartitions)

    def _layer(self):
        return {(self._name, i): self._task(k) for i, k in enumerate(self._partitions)}
```

Reply:
Or this:

```class Expr:
    def _layer(self):
        return {(self._name, i): self._task(k) for i, k in enumerate(getattr(self, "_partitions", range(self.npartitions)))}
```

Reply:
Okay - I was able to get this to work after I realized that we just need extra logic in `Fused._task` to deal with the fact that the index for one of the wrapped expressions may be different from the initial `index` argument.